Mathlib/LinearAlgebra/Span.lean

```diff
@@ -0,0 +1,1075 @@
+
+-- Porting note: TODO Erase this line. Needed because we don't have η for classes.
+attribute [-instance] Ring.toNonAssocRing
+
+-- Porting note: added the following line, fails to be inferred otherwise. Probably lean4#2074
+instance : Module K K := Semiring.toModule
+
+theorem span_singleton_sup_ker_eq_top (f : V →ₗ[K] K) {x : V} (hx : f x ≠ 0) :
+    (K ∙ x) ⊔ ker f = ⊤ :=
+  eq_top_iff.2 fun y _ =>
+    Submodule.mem_sup.2
+      ⟨(f y * (f x)⁻¹) • x, Submodule.mem_span_singleton.2 ⟨f y * (f x)⁻¹, rfl⟩,
+        ⟨y - (f y * (f x)⁻¹) • x, by
+          rw [LinearMap.mem_ker, f.map_sub, f.map_smul, smul_eq_mul, mul_assoc, inv_mul_cancel hx,
+            mul_one, sub_self],
+          by simp only [add_sub_cancel'_right]⟩⟩
+#align linear_map.span_singleton_sup_ker_eq_top LinearMap.span_singleton_sup_ker_eq_top
+
+variable (K V)
+
+theorem ker_toSpanSingleton {x : V} (h : x ≠ 0) : LinearMap.ker (toSpanSingleton K V x) = ⊥ := by
+  ext c; constructor
+  · intro hc
+    rw [Submodule.mem_bot]
+    rw [mem_ker] at hc
+    by_contra hc'
+    have : x = 0
+    calc
+      x = c⁻¹ • c • x := by rw [← mul_smul, inv_mul_cancel hc', one_smul]
+      _ = c⁻¹ • (toSpanSingleton K V x) c := rfl
+      _ = 0 := by rw [hc, smul_zero]
+    tauto
+  · rw [mem_ker, Submodule.mem_bot]
+    intro h
+    rw [h]
+    simp
+#align linear_map.ker_to_span_singleton LinearMap.ker_toSpanSingleton
+
+end Field
+
+end LinearMap
+
+open LinearMap
+
+namespace LinearEquiv
+
+section Field
+
+variable (K V) [Field K] [AddCommGroup V] [Module K V]
+
+-- Porting note: TODO Erase this line. Needed because we don't have η for classes.
+attribute [-instance] Ring.toNonAssocRing
+
+/-- Given a nonzero element `x` of a vector space `V` over a field `K`, the natural
+    map from `K` to the span of `x`, with invertibility check to consider it as an
+    isomorphism.-/
+noncomputable -- Porting note: added since `LinearEquiv.ofInjective` is noncomputable
+def toSpanNonzeroSingleton (x : V) (h : x ≠ 0) : K ≃ₗ[K] K ∙ x :=
+  LinearEquiv.trans
+    (LinearEquiv.ofInjective (LinearMap.toSpanSingleton K V x)
+      (ker_eq_bot.1 <| LinearMap.ker_toSpanSingleton K V h))
+    (LinearEquiv.ofEq (range $ toSpanSingleton K V x) (K ∙ x) (span_singleton_eq_range K V x).symm)
+#align linear_equiv.to_span_nonzero_singleton LinearEquiv.toSpanNonzeroSingleton
+
+theorem toSpanNonzeroSingleton_one (x : V) (h : x ≠ 0) :
+    LinearEquiv.toSpanNonzeroSingleton K V x h 1 =
+      (⟨x, Submodule.mem_span_singleton_self x⟩ : K ∙ x) := by
+  apply SetLike.coe_eq_coe.mp
+  have : ↑(toSpanNonzeroSingleton K V x h 1) = toSpanSingleton K V x 1 := rfl
+  rw [this, toSpanSingleton_one, Submodule.coe_mk]
+#align linear_equiv.to_span_nonzero_singleton_one LinearEquiv.toSpanNonzeroSingleton_one
+
+/-- Given a nonzero element `x` of a vector space `V` over a field `K`, the natural map
+    from the span of `x` to `K`.-/
+noncomputable -- Porting note: added since `toSpanNonzeroSingleton` is noncomputable
```

Comment:
Not sure this comment is needed; it's noncomputable in lean3 too.

Reply:
Ok, not sure why mathport didn't pick this up.